### PR TITLE
Extend API to support JOL IntelliJ Idea plugin

### PR DESCRIPTION
**Pull request by stokito**
[JOL Intellij Idea Plugin](https://github.com/stokito/IdeaJol)  internally uses JOL.
Internally in the [PsiClassAdapter](https://github.com/stokito/IdeaJol/blob/master/src/main/java/com/github/stokito/IdeaJol/PsiClassAdapter.java) it makes exactly the same as `org.openjdk.jol.info.ClassData#parse` but it creates the ClassData from Idea's AST/Psi tree instead of raw `Class`. Unfortunately to be able to do this we need to extend API a little bit.

Here is a short list of changes that I did:
1. Added an overloaded method ClassData.addSuperClassData(ClassData) so I can create an instance of `ClassData` 
2. Added `lossesInternal`, `lossesExternal` and `lossesTotal` fields to `ClassLayout` so they calculated only once on instance creation and then can be accessed multiple times via geters by JTable 
3. `FieldData.create()` internally evaluate if the field is contended but we do this based on Psi tree. So we need to open and make public the `FieldData` constructor.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Change must be properly reviewed

### Reviewers
 * [Aleksey Shipilev](https://openjdk.java.net/census#shade) (@shipilev - Committer) ⚠️ Review applies to 136d7743f4c575f01cf03161649492fe5ab3c59b

### Download
`$ git fetch https://git.openjdk.java.net/jol pull/4/head:pull/4`
`$ git checkout pull/4`

### CODE REVIEW

File: jol-core/src/test/java/org/openjdk/jol/info/ClassDataTest.java [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+    assertEquals(klass, classFields.get(3).hostClass());
+    assertEquals("[c3f1: int, c3f2: int, c3f3: int, c3f4: java.lang.Object]", classFields.toString());
+  }
+
+  @Test
+  public void fieldsForNonExisting() {
+    List<FieldData> classFields = classData.fieldsFor("Non existent class");
+    assertTrue(classFields.isEmpty());
+  }
+}
```
Comment (shipilev) [inline jol-core/src/test/java/org/openjdk/jol/info/ClassDataTest.java:93]: No newline.

File: jol-core/src/main/java/org/openjdk/jol/info/ClassLayout.java [hunk 2]
```diff
@@ -114,24 +127,33 @@ public static ClassLayout parseInstance(Object instance, Layouter layouter) {
      * @param instanceSize instance size
      * @param check        whether to check important invariants
      */
-    public ClassLayout(ClassData classData, SortedSet<FieldLayout> fields, int headerSize, long instanceSize, boolean check) {
-        this.classData = classData;
-        this.fields = fields;
-        this.headerSize = headerSize;
-        this.size = instanceSize;
+    public static ClassLayout createClassLayout(ClassData classData, SortedSet<FieldLayout> fields, int headerSize, long instanceSize, boolean check) {
```
Comment (shipilev) [inline jol-core/src/main/java/org/openjdk/jol/info/ClassLayout.java:130]: `ClassLayout.createClassLayout` seems tautological, maybe `ClassLayout.create`?

File: jol-core/src/main/java/org/openjdk/jol/info/ClassLayout.java [hunk 3]
```diff
@@ -114,24 +127,33 @@ public static ClassLayout parseInstance(Object instance, Layouter layouter) {
...
+        long lossesInternal = 0;
+        for (FieldLayout fieldLayout : fields) {
+            if (fieldLayout.offset() > nextFree) {
+                long fieldLayoutSize = fieldLayout.offset() - nextFree;
+                lossesInternal += fieldLayoutSize;
+            }
+            nextFree = fieldLayout.offset() + fieldLayout.size();
+        }
+        long lossesExternal = instanceSize != nextFree ? instanceSize - nextFree : 0;
+        long lossesTotal = lossesInternal + lossesExternal;
```
Comment (shipilev) [inline jol-core/src/main/java/org/openjdk/jol/info/ClassLayout.java:145]: I think this whole block would be more readable if we shorten the variable names.

```
        long next = headerSize;
        long internal = 0;
        for (FieldLayout fl : fields) {
            if (fl.offset() > nextFree) {
                internal += fl.offset() - nextFree;
            }
            next = fl.offset() + fl.size();
        }
        long external = (instanceSize != next) ? (instanceSize - next) : 0;
        long total = internal + external;
```

File: jol-core/src/test/java/org/openjdk/jol/info/ClassDataTest.java (hunk 1 above)
Comment (shipilev) [inline jol-core/src/test/java/org/openjdk/jol/info/ClassDataTest.java:93]: Add a newline at the end of file.

File: jol-core/src/main/java/org/openjdk/jol/info/ClassLayout.java (hunk 2 above)
Comment (shipilev) [reply jol-core/src/main/java/org/openjdk/jol/info/ClassLayout.java:130]: This also matches e.g. `FieldData.create`.

File: jol-core/src/main/java/org/openjdk/jol/info/FieldData.java [hunk 4]
```diff
@@ -74,7 +74,7 @@ public static FieldData parse(Field field) {
     private final boolean isContended;
     private final String contendedGroup;
 
-    private FieldData(Field refField, String hostKlass, String fieldName, String fieldType,
+    public FieldData(Field refField, String hostKlass, String fieldName, String fieldType,
```
Comment (shipilev) [inline jol-core/src/main/java/org/openjdk/jol/info/FieldData.java:77]: Don't you want to add the overloaded `FieldData.create` instead?

**Review by shipilev — CHANGES_REQUESTED** [review]
Looks sensible. Please rename the PR to e.g. "Extend API to support JOL IntelliJ Idea plugin"?

File: jol-core/src/main/java/org/openjdk/jol/info/ClassLayout.java (hunk 3 above)
Comment (stokito) [reply jol-core/src/main/java/org/openjdk/jol/info/ClassLayout.java:145]: ok, I just kept original variable names for more clear history

**Comment by stokito** [comment]
I fixed all your review comments but I wondered that I not listed as signed OCA. I'll sign it again but this may take for weeks

**Review by shipilev — APPROVED** [review]
This looks fine. Again, I'll integrate once OCA and other integration blockers resolve and bots allow me to integrate.

**Comment by shipilev** [comment]
@stokito, could you please merge/rebase off the current `master`? It includes the automatically-triggered workflow with pre-integration checks.

**Pull request closed without merging**